<a href="https://colab.research.google.com/github/manuelcampagnolo/exemplos_python/blob/main/rf_embeddings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Saving embedding_T29TNE_2018_2021_all.csv to embedding_T29TNE_2018_2021_all.csv
User uploaded file "embedding_T29TNE_2018_2021_all.csv" with length 15629844 bytes


banda_0   banda_1   banda_2   banda_3   banda_4   banda_5   banda_6  \
0  0.130165  0.024606 -0.093564  0.017778  0.141730 -0.071111  0.038447   
1  0.141730  0.041584 -0.084214  0.006151  0.130165 -0.062991  0.038447   
2  0.147697  0.022207 -0.071111 -0.000062  0.130165 -0.002215  0.012057   
3  0.119093  0.019931 -0.098424  0.012057  0.130165 -0.079723  0.035433   
4  0.172795  0.008858 -0.055363  0.017778  0.130165 -0.008858  0.019931   

    banda_7   banda_8   banda_9  ...  banda_59  banda_60  banda_61  banda_62  \
0  0.019931  0.251965 -0.017778  ... -0.251965  0.192910  0.066990 -0.022207   
1  0.012057  0.228897 -0.029773  ... -0.267958  0.179377  0.079723 -0.007443   
2 -0.022207  0.236463 -0.032541  ... -0.259900  0.244152  0.003937  0.010396   
3  0.029773  0.251965 -0.012057  ... -0.267958  0.199862  0.059116 -0.029773   
4  0.000554  0.251965 -0.032541  ... -0.244152  0.244152  0.017778 -0.000062   

   banda_63         x          y  label  poligono_id   ano  
0 -0.075356  599025.0  4450235.0      1          116  2018  
1 -0.079723  599015.0  4450205.0      1          116  2018  
2 -0.098424  598945.0  4450055.0      1          116  2018  
3 -0.055363  599045.0  4450245.0      1          116  2018  
4 -0.098424  598935.0  4450045.0      1          116  2018  

[5 rows x 69 columns]

# Task
Analyze the provided data to predict the 'label' column using a Random Forest classifier. Exclude the columns 'x', 'y', and 'ano' from the features. Perform a 5-fold cross-validation using GroupKFold based on the 'poligono_id' column. Report the cross-validation scores.

## Prepare data

### Subtask:
Separate the features (all columns except 'label', 'x', 'y', and 'ano') and the target variable ('label'). Also, create a variable for the groups ('poligono_id').


**Reasoning**:
Separate the features, target, and groups based on the instructions.



In [4]:
feature_columns = [col for col in df.columns if col not in ['label', 'x', 'y', 'ano']]
X = df[feature_columns]
y = df['label']
groups = df['poligono_id']

## Initialize random forest

### Subtask:
Set up a Random Forest classifier.


**Reasoning**:
Import the necessary class and instantiate the Random Forest classifier as instructed.



In [5]:
from sklearn.ensemble import RandomForestClassifier

rf_classifier = RandomForestClassifier(random_state=42)

## Set up groupkfold

### Subtask:
Configure GroupKFold for 5 splits using the 'poligono_id' groups.


**Reasoning**:
Configure GroupKFold for 5 splits using the 'poligono_id' groups.



In [6]:
from sklearn.model_selection import GroupKFold

group_kfold = GroupKFold(n_splits=5)

## Perform cross-validation

### Subtask:
Use `cross_val_score` with the Random Forest model, the feature data, the target variable, the groups, and the GroupKFold strategy.


**Reasoning**:
Import the necessary function and perform the cross-validation.



In [7]:
from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(rf_classifier, X, y, groups=groups, cv=group_kfold)

## Evaluate results

### Subtask:
Print the cross-validation scores.


**Reasoning**:
Print the cross-validation scores stored in the `cv_scores` variable to fulfill the subtask requirement.



In [8]:
print("Cross-validation scores:", cv_scores)

Cross-validation scores: [0.9156 0.8608 0.8784 0.8496 0.8712]


## Summary:

### Data Analysis Key Findings

*   The feature set used for prediction included all columns except 'label', 'x', 'y', and 'ano'.
*   A Random Forest Classifier with a `random_state` of 42 was used for the prediction.
*   A 5-fold cross-validation was performed using GroupKFold based on the 'poligono\_id' column.
*   The cross-validation scores obtained were: \[0.9156 0.8608 0.8784 0.8496 0.8712].

### Insights or Next Steps

*   The cross-validation scores indicate reasonable performance across different groups, with scores ranging from approximately 0.85 to 0.92.
*   Further analysis could involve calculating the mean and standard deviation of the cross-validation scores to get an overall performance metric and its variability.


## Calculate average confusion matrix

### Subtask:
Perform cross-validation manually to calculate and average the confusion matrices from each fold.

**Reasoning**:
Manually iterate through the splits generated by GroupKFold, train the model on the training fold, predict on the test fold, calculate the confusion matrix, and accumulate them to calculate the average.

In [9]:
from sklearn.metrics import confusion_matrix
import numpy as np

# Initialize a variable to accumulate confusion matrices
average_confusion_matrix = np.zeros((len(np.unique(y)), len(np.unique(y))))

# Perform cross-validation manually
for fold, (train_index, test_index) in enumerate(group_kfold.split(X, y, groups)):
    print(f"Processing fold {fold+1}/5")
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    # Train the classifier
    rf_classifier.fit(X_train, y_train)

    # Predict on the test data
    y_pred = rf_classifier.predict(X_test)

    # Calculate confusion matrix for the current fold
    cm = confusion_matrix(y_test, y_pred)

    # Accumulate the confusion matrix
    average_confusion_matrix += cm

# Calculate the average confusion matrix by dividing by the number of folds
average_confusion_matrix = average_confusion_matrix / group_kfold.get_n_splits()

print("\nAverage Confusion Matrix across 5 folds:")
print(average_confusion_matrix)

Processing fold 1/5
Processing fold 2/5
Processing fold 3/5
Processing fold 4/5
Processing fold 5/5

Average Confusion Matrix across 5 folds:
[[ 226.   274. ]
 [  38.2 1961.8]]


In [10]:
label_distribution = df['label'].value_counts()
print("Distribution of 'label' column:")
print(label_distribution)

Distribution of 'label' column:
label
1    10000
0     2500
Name: count, dtype: int64
